hub/src/server/drivers/GcDriver.ts

```diff
@@ -222,7 +222,71 @@ class GcDriver implements DriverModel, DriverModelTestMethods {
         ` ${filename} in bucket ${this.bucket}: ${error}`)
     }
   }
-  
+
+  async performStat(args: PerformStatArgs): Promise<StatResult> {
+    if (!GcDriver.isPathValid(args.path)) {
+      throw new BadPathError('Invalid Path')
+    }
+    const filename = `${args.storageTopLevel}/${args.path}`
+    const bucketFile = this.storage
+      .bucket(this.bucket)
+      .file(filename)
+    try {
+      const [metadataResult] = await bucketFile.getMetadata()
+      const lastModified = Math.round(new Date(metadataResult.updated).getTime() / 1000)
+      const result: StatResult = {
+        exists: true,
+        contentType: metadataResult.contentType,
+        contentLength: parseInt(metadataResult.size),
+        lastModifiedDate: lastModified
+      }
+      return result
+    } catch (error) {
+      if (error.code === 404) {
+        const result: StatResult = {
+          exists: false
+        }
+        return result
+      }
+      /* istanbul ignore next */
+      logger.error(`failed to stat ${filename} in bucket ${this.bucket}`)
+      /* istanbul ignore next */
+      throw new Error('Google cloud storage failure: failed to stat ' +
+        ` ${filename} in bucket ${this.bucket}: ${error}`)
+    }
+  }
+
+  async performRename(args: PerformRenameArgs): Promise<void> {
```

Comment:
On some of the `performRename` implementations, we follow up the rename with a `performStat()`-like operation on the new path to verify that the rename went through.  I'm guessing this is because rename isn't atomic (or even available as a first-class operation) on some storage providers?  If so, can we explicitly confirm whether or not each of these backends does renames atomically, and add a comment that points to the relevant documentation indicating as such?

Reply:
S3 and Azure APIs do not support a rename/move operation. The recommended approach is to perform a copy then delete. Their CLI tools and other language libraries (like Java and C#) that implement anything like a `move` all do this `copy & delete` behavior under the hood, on the client. 

Integration tests for all providers validating `copy & delete` === `rename` for relevant drivers implemented in https://github.com/blockstack/gaia/pull/246/commits/f77e09db3ebfc8fb30f5d49c93a2c189813f3671

For concurrency issues -- a mutex around file ends points (`address/filename`) will be implemented around the http handler layer. Multi-hub load balancers will need to partition requests based on something like path prefix or hashing, rather than round-robin.

Reply:
Gotcha.  Thanks for addressing this!